### ISSUE TITLE
Data Loader, CLI upsert "External_ID__C not specified"

### ISSUE THREAD

**Issue by chimpanzee818** [forcedotcom/dataloader#717]
@ashitsalesforce 

my issue is similar to below mentioned issue. 

"https://developer.salesforce.com/forums/?id=906F0000000kBKOIA2" 

having everything similar and error too.

please response 
and my version of Data Loader is 54 

thanks in advance,

**Comment by ashitsalesforce** [comment]
Hi @chimpanzee818 , I followed the details provided in the [issue](https://developer.salesforce.com/forums/?id=906F0000000kBKOIA2) using Data Loader 58.0.3 but couldn't reproduce it. Here is the result in success csv file:
```
"ID","KONTOFÜHRER-NR.","KONTOFÜHRER-NAME","SAP-NAME","TELEFONNUMMER","EMAIL-ADRESSE","NW","STATUS"
"0034E0000XXXXXXXXX","01","M Mustermann","NW1234","023314612345","test.test@test.com","NW","Item Created"
```

Provide the same information given in the issue - relevant details from process-conf.xml, content of the CSV file, content of the sdl file, the batch command you execute, and its error output - if you want us to help you troubleshoot it further.

**Comment by chimpanzee818** [comment]
hy @ashitsalesforce 

****my script**** 

cd /pathway/
for fileName in *.csv; do
        java -cp dataloader-55.0.1-uber.jar -Dsalesforce.config.dir="conf/" com.salesforce.dataloader.process.DataLoaderRunner run.mode=batch process.name=loadData dataAccess.name="$fileName"

#       mv -f $fileName archive/$fileName
done

**config**
sfdc.endpoint=https://test.salesforce.com
sfdc.username=user@fullsbox
sfdc.password=encrypted pwd
process.encryptionKeyFile=/path/.dataloader/dataLoader.key


**process conf .xml**
<!DOCTYPE beans PUBLIC "-//SPRING//DTD BEAN//EN" "http://www.springframework.org/dtd/spring-beans.dtd">
<beans>
    <bean id="loadData" class="com.salesforce.dataloader.process.ProcessRunner" scope="prototype">
        <description>Load data into Salesforce.</description>
        <property name="name" value="loadData"/>
        <property name="configOverrideMap">
            <map>
                <entry key="sfdc.entity" value="object"/>
                <entry key="process.operation" value="upsert"/>
                <entry key="sfdc.externalIdField" value="DHDKEY__c"/>
                <entry key="process.mappingFile" value="conf/mapping.sdl"/>
                <entry key="dataAccess.type" value="csvRead"/>
                <entry key="loader.csvComma" value="false"/>
              </map>
        </property>
    </bean>
</beans>

**mapping.sdl**
#CSV_COLUMN_HEADER=SALESORCE_API_FIELD_NAME__C
filekey=DHDKEY__c
sid=Contact__r\:DSI__c
aidyr=aid_Year__c
sessid=Session__r\:Session_Unique_Code__c

**error**
- Mapping for field filekey,sid,aidyr,term,sessid,end will be ignored since destination column is empty
2023-07-07 01:50:42,777 ERROR [loadData] client.PartnerClient processResult (PartnerClient.java:459) - Errors were found on item0
2023-07-07 01:50:42,777 ERROR [loadData] client.PartnerClient processResult (PartnerClient.java:461) - Error code is:MISSING_ARGUMENT
2023-07-07 01:50:42,777 ERROR [loadData] client.PartnerClient processResult (PartnerClient.java:462) - Error message:DHDKEY__c not specified
2023-07-07 01:50:42,779 INFO  [loadData] progress.NihilistProgressAdapter setSubTask (NihilistProgressAdapter.java:78) - Processed 1 of 1 total records. Rate: 0 records per hour. Estimated time to complete: 153,722,867,280,912 minutes and 55 seconds.  There are 0 successes and 1 errors.
2023-07-07 01:50:42,779 INFO  [loadData] progress.NihilistProgressAdapter doneSuccess (NihilistProgressAdapter.java:63) - The operation has fully completed.  There were 0 successful upserts and 1 errors.

**CSV**
filekey,sid,aidyr,term,sessid,end
D41167  ,D411,2223,5,202310,end


these are the total details

**Comment by ashitsalesforce** [comment]
Hi @chimpanzee818 ,

What entity are you upserting? I did not find a standard entity named "object" in my org.

**Comment by chimpanzee818** [comment]
@ashitsalesforce 

Basically I am using custom object not standard obj

**Comment by ashitsalesforce** [comment]
If I get it right, you have a custom object named "cobject", and the line in process-conf.xml is `<entry key="sfdc.entity" value="cobject__c"/>`. Is that right?

Also, I am assuming you have a custom field `DSI` on `Contact` object that's an external ID field. What about `Session` object? Is it a custom object?

Lastly, try with `sid=Contact:DSI__c` and `sessid=Session__c:Session_Unique_Code__c` (assuming that Session is a custom object for your org as I am not able to reproduce the issue.

**Comment by chimpanzee818** [comment]
@ashitsalesforce 

I am using custom object

**Comment by chimpanzee818** [comment]
@ashitsalesforce 

Yes u r right that is custom obj

Session is also custom object nd dsi relation with contact is correct

U r telling me try the thing in mapping.sdl file ?

Instead of 
`sid=Contact__r:DSI__c` i hv to go `sid=Contact:DSI__c` 

 but here contact__c field nm on cobject (custom obj ) having custom relationship between contact nd cobject

Then instead of using contact__r use contact

**Comment by ashitsalesforce** [comment]
Hi @chimpanzee818 , your current mapping.sdl settings are right. I took the following steps:
- Created `object__c` custom object with fields `DHDKEY__c` (type: text, external ID set to true) and `aidYear__c` (type: number).
- Created `session__c` custom object with field `Session_Unique_Code__c` (type: number, external ID set to true).
- Added field `DSI__c` (type: number, external ID set to true) to `Contact` standard object.
- Set `DSI__c` value to `D411` on an existing Contact record.
- Create a new `Session` record with `Session_Unique_Code__c` set to `202310`.
- Copied your mapping.sdl and csv content as-is and ran data loader v58.0.3 in batch mode twice, first time to create a new `object` record through upsert operation, second time to update the same record through upsert operation. Both of these runs executed successfully.

Going again through your process-conf.xml snippet, I found the entry `<entry key="loader.csvComma" value="false"/>` that did not exist in my process-conf.xml. I added that entry and was able to reproduce your issue. So, the fix is to delete that entry from process-conf.xml.
